account/account.go

```diff
@@ -45,58 +41,335 @@ func SignBytes(chainID string, o Signable) []byte {
+type Addressable interface {
+	// Get the 20 byte EVM address of this account
+	Address() Address
+	// Public key from which the Address is derived
+	PublicKey() PublicKey
+}
 
-// Account resides in the application state, and is mutated by transactions
-// on the blockchain.
-// Serialized by wire.[read|write]Reflect
-type Account struct {
-	Address     []byte        `json:"address"`
-	PubKey      crypto.PubKey `json:"pub_key"`
-	Sequence    int           `json:"sequence"`
-	Balance     int64         `json:"balance"`
-	Code        []byte        `json:"code"`         // VM code
-	StorageRoot []byte        `json:"storage_root"` // VM storage merkle root.
+// The default immutable interface to an account
+type Account interface {
+	Addressable
+	// The value held by this account in terms of the chain-native token
+	Balance() uint64
+	// The EVM byte code held by this account (or equivalently, this contract)
+	Code() Bytecode
+	// The sequence number (or nonce) of this account, incremented each time a mutation of the
+	// Account is persisted to the blockchain state
+	Sequence() uint64
+	// The hash of all the values in this accounts storage (typically the root of some subtree
+	// in the merkle tree of global storage state)
+	StorageRoot() []byte
+	// The permission flags and roles for this account
+	Permissions() ptypes.AccountPermissions
+	// Obtain a deterministic serialisation of this account
+	// (i.e. update order and Go runtime independent)
+	Encode() []byte
+}
 
-	Permissions ptypes.AccountPermissions `json:"permissions"`
+type MutableAccount interface {
+	Account
+	// Set public key (needed for lazy initialisation), should also set the dependent address
+	SetPublicKey(pubKey PublicKey) MutableAccount
+	// Subtract amount from account balance (will panic if amount is greater than balance)
+	SubtractFromBalance(amount uint64) MutableAccount
+	// Add amount to balance (will panic if amount plus balance is a uint64 overflow)
+	AddToBalance(amount uint64) MutableAccount
+	// Set EVM byte code associated with account
+	SetCode(code []byte) MutableAccount
+	// Increment Sequence number by 1 (capturing the current Sequence number as the index for any pending mutations)
+	IncSequence() MutableAccount
+	// Set the storage root hash
+	SetStorageRoot(storageRoot []byte) MutableAccount
+	// Set account permissions
+	SetPermissions(permissions ptypes.AccountPermissions) MutableAccount
+	// Get a pointer this account's AccountPermissions in order to mutate them
+	MutablePermissions() *ptypes.AccountPermissions
+	// Create a complete copy of this MutableAccount that is itself mutable
+	Copy() MutableAccount
 }
 
-func (acc *Account) Copy() *Account {
-	accCopy := *acc
-	return &accCopy
+// -------------------------------------------------
+// ConcreteAccount
+
+// ConcreteAccount is the canonical serialisation and bash-in-place object for an Account
+type ConcreteAccount struct {
```

Comment:
Same for other types following the `T` -> `ConcreteT` pattern

Reply:
I'm not in love with the name either. `BurrowAccount` is a bit redundant since 'Burrow' is in the namespace already, also `Account` is the real Burrow account. The ugly name is probably a symptom of the ugly concept... Open to other suggestions.